core/src/main/java/org/apache/brooklyn/entity/group/DynamicClusterImpl.java

```diff
@@ -799,6 +795,14 @@ protected Entity replaceMember(Entity member, @Nullable Location memberLoc, Map<
     /** <strong>Note</strong> for sub-classes; this method can be called while synchronized on {@link #mutex}. */
     protected Collection<Entity> grow(int delta) {
         Preconditions.checkArgument(delta > 0, "Must call grow with positive delta.");
+        Integer maxSize = config().get(MAX_SIZE);
+        if (maxSize != null) {
+            final int desiredSize = getCurrentSize() + delta;
+            if (desiredSize > maxSize) {
+                throw new Resizable.InsufficientCapacityException(
```

Comment:
Should we throw if we'd be willing to grow a bit? e.g. if we're current size 1 and asked to resize by delta of 10, but max size is 10 then should we resize to 10 or should we fail? I think we should only fail if `getCurrentSize() == maxSize`

Reply:
I think throwing here is the right behaviour @aledsage: we cannot grow more that the maximum size allowed. It also preserves backward compatibility.

However, the idea of growing until we reach the max size, regardless of exceeding it is an interesting idea. Maybe this could be another effector, or some parameter to the `resize` to allow it?

Reply:
@tbouron I don't think we want another effector or parameter as we wouldn't get the "desired default behaviour" in our existing usage, e.g. from the auto-scaler policy (assuming we can agree what that desired behaviour is).

The reason I think we should resize it to as big as we're allowed is that I believe that is what will work best with our existing usage. For example, if an auto-scaler thinks we need to go to size 11 but we're only allowed to go to 10, then leaving it at the existing much smaller size seems very wrong. The auto-scaler, as currently written, would never discover that if it asked for 10 then it would get it; the auto-scaler would just keep asking for the bigger value that it thinks is necessary.

I admit it will slightly confuse users if they say "resize to 11" and it only goes to 10, without anything but a warn message in the log. However, at least it's done something; and if they call it again with "11" then they'll get their exception.

Reply:
I took a fairly strict interpretation of DynamicCluster's contract. If it's not possible to fulfil a request for a particular cluster size, or *n* new members, in full then I thought it should throw. The consumer should be smarter, expect an `InsufficientCapacityException` and ask for less. In the case of an auto-scaler I'd expect it to make a second attempt to resize.

If the arguments to resize and resizeByDelta are suggestions rather than requirements then best-effort is acceptable.